splash-lp-indexer/src/position_db/mature_events.rs

```diff
@@ -125,155 +157,190 @@ impl MatureEvents for PositionDB {
     let mut aggregated_events: HashMap<PoolId, EventsByPool> = HashMap::new();
     for event in events {
         let event_pid = event.pool_id();
         match aggregated_events.entry(event_pid) {
             Entry::Vacant(entry) => {
                 let mut new_frame = EventsByPool::new();
-                new_frame.apply_event(event);
+                let lp_supply = new_frame.apply_event(event);
+                if let Some(lp_supply) = lp_supply {
+                    set_pool_lp_supply(tx, cfs.pool_lq, event_pid, lp_supply);
+                }
                 entry.insert(new_frame);
             }
             Entry::Occupied(mut entry) => {
-                entry.get_mut().apply_event(event);
+                let lp_supply = entry.get_mut().apply_event(event);
+                if let Some(lp_supply) = lp_supply {
+                    set_pool_lp_supply(tx, cfs.pool_lq, event_pid, lp_supply);
+                }
             }
         };
     }
     aggregated_events
 }
 
-/// All events that happened for a single account in a single pool in a single slot.
-#[derive(Debug)]
-struct EventsByAccount {
-    position_events: Vec<PositionEvent>,
-}
-
-impl EventsByAccount {
-    fn new() -> Self {
-        Self {
-            position_events: vec![],
+/// Returns the latest account positions that have positive share for each account.
+fn get_latest_account_positions(
+    db: &Arc<TransactionDB>,
+    cf: &ColumnFamily,
+    pool_key: Vec<u8>,
+) -> HashMap<Credential, (Epoch, AccountPosition)> {
+    let iter_positions = get_range_iterator(db, cf, pool_key).filter_map(|e| match e {
+        Ok((key, value)) => {
+            let (_, account_cred, position_epoch) = parse_position_key(key.to_vec())?;
+            let current_position = rmp_serde::from_slice::<AccountPosition>(&value).ok()?;
+            Some((account_cred, position_epoch, current_position))
+        }
+        Err(_) => None,
+    });
+    let mut latest_positions: HashMap<Credential, (Epoch, AccountPosition)> = HashMap::new();
+    for (account_cred, position_epoch, current_position) in iter_positions {
+        if let Some((latest_epoch, _)) = latest_positions.get(&account_cred) {
+            if position_epoch > *latest_epoch && !current_position.is_currently_zero_share() {
+                latest_positions.insert(account_cred, (position_epoch, current_position));
+            }
+        } else {
+            latest_positions.insert(account_cred, (position_epoch, current_position));
         }
     }
-    fn apply_event(&mut self, event: PositionEvent) -> Option<u64> {
-        let lp_supply = event.lp_supply();
-        self.position_events.push(event);
-        Some(lp_supply)
-    }
+    latest_positions
 }
 
 /// All events that happened for a single pool in a single slot.
 #[derive(Debug)]
 struct EventsByPool {
     gauge_events: Vec<GaugeWeighted>,
-    account_frames: HashMap<Credential, EventsByAccount>,
+    account_frames: Vec<PositionEvent>,
```

Comment:
Using `HashMap`, we lose the ordering of events which will lead to incorrect reward quantities from `lp-indexer`.